### PR TITLE
#38 add support for pyspark trees in shap explainer

### PR DESCRIPTION
**Pull request by QuentinAmbard**
This pull request add support for pyspark Decision Trees (Random Forest and GBT) in the explainer.
It doesn't use spark to explain the model, big dataset still need to be reduced and converted as panda DF in order to run the explainer.

**Limitations:**

- Categorical split aren't supported, I haven't seen this feature being supported in SHAP, if it is I'd be happy to add it but I don't see a simple way to add it
- Variance impurity isn't supported
- the .predict() function doesn't support prediction with spark

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
Looks nice!

File: tests/explainers/test_tree.py [hunk 1]
```diff
@@ -187,6 +187,40 @@ def test_xgboost_mixed_types():
...
+    dt = DecisionTreeClassifier(labelCol="label", featuresCol="features")
+    model = dt.fit(iris)
+    explainer = shap.TreeExplainer(model)
+    X = pd.DataFrame(data=iris_sk.data, columns=iris_sk.feature_names) # pylint: disable=E1101
+
+    shap_values = explainer.shap_values(X)
+    spark.stop()
+
+#_validate_shap_values(model, x_test)
+test_pyspark_decision_tree()
```
Comment (slundberg) [inline tests/explainers/test_tree.py]: you don't need to explicitly call the test here

File: tests/explainers/test_tree.py [hunk 2]
```diff
@@ -187,6 +187,40 @@ def test_xgboost_mixed_types():
...
+    iris = spark.createDataFrame(iris, col)
+    iris = VectorAssembler(inputCols=col[:-1],outputCol="features").transform(iris)
+    iris = StringIndexer(inputCol="type", outputCol="label").fit(iris).transform(iris)
+
+    dt = DecisionTreeClassifier(labelCol="label", featuresCol="features")
+    model = dt.fit(iris)
+    explainer = shap.TreeExplainer(model)
+    X = pd.DataFrame(data=iris_sk.data, columns=iris_sk.feature_names) # pylint: disable=E1101
+
+    shap_values = explainer.shap_values(X)
```
Comment (slundberg) [inline tests/explainers/test_tree.py:218]: Can you add a quick check to ensure that the sum of the returned SHAP values (plus explainer.expected_value) equals the output of the model?

**Comment by slundberg** [comment]
Thanks so much! looks good. I just a left a few small comments.

**Comment by slundberg** [comment]
Thanks!

**Comment by QuentinAmbard** [comment]
I realised I only added support for Classification in this PR, I'll do another for DecisionTreeRegressionModel soon

**Comment by ksj3421** [comment]
Thank you for your contribute. I also want to use shap with spark model. Currently i am using multiclass randomforest spark model , is there any way to support predict function also work for spark what you mentoned in limitation. I cant get my rawprediction value with expected value and sum of specific class of shap value. Want to solve this problem!! Thank you in advance

**Comment by QuentinAmbard** [comment]
@ksj3421 why not running it using the spark ml api directly? It shouldn't change anything, use your spark model to run your prediction in a spark job and then analyse the predictions with shap & spark if needed. Sean published a great example here: https://databricks.com/blog/2019/06/17/detecting-bias-with-shap.html

**Comment by ksj3421** [comment]
Thank you for your comment! But i still have question.. What i meant was for rf multiclass model, my rawprediction values (from spark prediction)  and  force plot's last expect value(calculated from shap value) has so big gap. For example if there is 5 classes, i can get 5 each rawprediction and shapley added with expected value!

**Comment by allard-jeff** [comment]
@QuentinAmbard 
Just ran across this great news! Are there any demos on how to use? Does one use SHAP as is, just referencing a pyspark model - that simple?

**Comment by QuentinAmbard** [comment]
@allardAmwayDS yes you can just pass the pyspark model in the explainer, simple as that.
Example in the unit tests here:
https://github.com/QuentinAmbard/shap/blob/%23866_spark_regressor_support/tests/explainers/test_tree.py#L190
If you want to distribute shap computation, it's quite easy to do with spark using a pandas udf (just put the computation of the shap_values in your Pandas Scalar iterator UDFs and return them as extra column) .

Once https://github.com/slundberg/shap/issues/866 is merged all pyspark tree algorithm should be supported

**Comment by allard-jeff** [comment]
@QuentinAmbard  I installed the current master branch and when I try to run this:
`shap_values = explainer.shap_values(X)`

I get the following:
`NotImplementedError: Predict with pyspark isn't implemented`

Am I missing something - this is a required step of the Shap algorithm.

**Comment by QuentinAmbard** [comment]
@allardAmwayDS  I assume that's because you send some data to the tree explainer, it's only going to work with tree_path_dependent and with background dataset being None (and use the information from the model instead).
Supporting spark prediction adds a bit of complexity I haven't implemented it yet. Feel free to open a ticket if you need it.

**Comment by allard-jeff** [comment]
@QuentinAmbard , I was following the unit tests (I thought) linked to above. It appears you are passing in a test set.

```
    for classifier in classifiers:
        model = classifier.fit(iris)
        explainer = shap.TreeExplainer(model)
        X = pd.DataFrame(data=iris_sk.data, columns=iris_sk.feature_names)[:100] # pylint: disable=E1101

        shap_values = explainer.shap_values(X)

```

1) If not this, how to use?
2) What is the implication of not having predict coded for pyspark in terms of the Shap values?

**Comment by QuentinAmbard** [comment]
No I'm not, I'm talking about background data passed in the TreeExplainer, which I don't do in the unit test: shap.TreeExplainer(model, data=None)
explainer.shap_values(X) doesn't have to run predictions in this case as it's using the model information.

**Comment by allard-jeff** [comment]
Its the last bit (explainer.shap_values(X)) where I got the error. I did not explicitly set data=None in shap.TreeExplainer(model) though. Perhaps thats the issue.

**Comment by allard-jeff** [comment]
Should we expect a degrade in the "quality" of the Shap values with this restriction?

**Comment by QuentinAmbard** [comment]
Please share your full example in a separate issue, it should be working.
It hasn't "quality" issue, you can only use the 'tree_path_dependent' feature_dependence, which is the default.

**Comment by allard-jeff** [comment]
Trying to reproduce the unit tests, now I am finding that if I install from git (clone, python setup.py install) I get this error. So not sure.

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-13-2ceacf55435b> in <module>
     22 
     23 model = classifier.fit(iris)
---> 24 explainer = shap.TreeExplainer(model)
     25 X = pd.DataFrame(data=iris_sk.data, columns=iris_sk.feature_names)[:100] # pylint: disable=E1101
     26 X.head()

AttributeError: module 'shap' has no attribute 'TreeExplainer'

**Comment by joshTheECE** [comment]
Hi @QuentinAmbard I'm interested to use this implementation for the RandomForestClassifier we used using Spark. Our dataframe though is a Spark Dataframe. How are we going to provide "prediction explanation" for each row? Can you help provide a sample? In your sample, you are using a pandas dataframe but our data is too huge to fit in memory. Thank you in advance.

**Comment by QuentinAmbard** [comment]
@joshTheECE you just have to do it in a [pandas udf in spark](https://docs.databricks.com/spark/latest/spark-sql/udf-python-pandas.html#iterator-of-series-to-iterator-of-series-udf). 
In the udf, you load the explainer, and then iterate over the pandas batch. It'll score everything & work at infinite scale

**Pull request merged**